Refine Endpoint selection for multi-cluster Service

luolanzone opened the pull request:
When the Endpoint of Multi-cluster Service is a local Service ClusterIP, refine the action to let it go to the corresponding exported Service's group to do final Endpoint selection. This can avoid the case that the traffic goes out of antrea-gw0 and goes back to OVS again when a local Pod is trying to access a MC Service but a local Service's Endpoint is selected.

Resolve #4431 
Signed-off-by: Lan Luo <luola@vmware.com>

hongliangl commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2414,14 +2416,19 @@ func (f *featureService) serviceEndpointGroup(groupID binding.GroupIDType, withS
 		endpointIP := net.ParseIP(endpoint.IP())
 		portVal := util.PortToUint16(endpointPort)
 		ipProtocol := getIPProtocol(endpointIP)
-
```
Revert this

hongliangl replied:
Revert this.

hongliangl commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2942,6 +2949,15 @@ func (f *featureService) gatewaySNATFlows() []binding.Flow {
 }
 
+func (f *featureService) isLocalServiceClusterIP(ip net.IP) bool {
+	for _, cidr := range f.serviceCIDRs {
```
`serviceCIDRs` can be only used when AntreaProxy is disabled. We need another way to get the Service CIDRs.

luolanzone replied:
There is a way to find serviceCIDR automatically in MC controller, but I feel we should have one single source of truth for agent part. I believe reusing the `serviceCIDRs` config item and refine the comment is one way, not sure if there is any other way to share this kind of information between agents.

luolanzone replied:
I added a new field `serviceCIDR` under `multicluster` config section.

wenyingd commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2414,14 +2416,19 @@ func (f *featureService) serviceEndpointGroup(groupID binding.GroupIDType, withS
-				LoadToRegField(EndpointPortField, uint32(portVal)).
-				ResubmitToTable(resubmitTableID).
-				Done()
+			if exportedSvcGroupID != 0 && f.isLocalServiceClusterIP(endpointIP) {
```
No need to check if the endpointIP is local cluster Service IP in openflow layer or not, we can check it proxier, and in openflow layer just leverage the expected groupID in the buckets as long as it is valid?

luolanzone commented:
Hi @jianjuns @wenyingd @hongliangl I added a new config field 'serviceCIDR' under multicluster config section in this PR so that antrea-agent can check if Endpoint is local Service's ClusterIP or not. Another way to figure out the `serviceCIDR` of a cluster is to do like antrea-mc-controller to get the CIDR automatically, but it requires extra RBAC to allow antrea-agent to have Service creation capability.
Personally, I would prefer to let antrea-agent to figure out Service CIDR automatically which is more user-friendly way. Let me know if you have any suggestion. Thanks.

jianjuns commented:
Do we have a better way to handle this rather than checking against `serviceCIDR`? AntreaProxy already knows all local ClusterIPs right. @hongliangl

hongliangl commented:
> Do we have a better way to handle this rather than checking against `serviceCIDR`? AntreaProxy already knows all local ClusterIPs right. @hongliangl

Yes, Antrea Proxy already knows a CIDR that can contain all existing ClusterIPs, but it is located in pkg/agent/route, like the IPv4 Service CIDR https://github.com/antrea-io/antrea/blob/114e14f5988a7b730fea7757f12fa68bd247feb3/pkg/agent/route/route_linux.go#L122.

I think the Service CIDR can be obtained by adding a new public method to this interface https://github.com/antrea-io/antrea/blob/114e14f5988a7b730fea7757f12fa68bd247feb3/pkg/agent/route/interfaces.go#L25.

luolanzone commented:
@hongliangl I checked the codes, it seems Antrea proxy will only calculate the ServiceCIDR when proxyAll is enabled. Do you know how does AntreaProxy know what's the ServiceCIDR when proxyAll is disabled?

luolanzone commented:
@jianjuns  I double checked with hongliang about the current ServiceCIDR implementation on proxy. Antrea proxy will calculate ServiceCIDR only when proxyAll is enabled. So Hongliang's suggestion to expose ServiceCIDR from route pkg can work only when we enable proxyAll with multicluster.  I feel this is not applicable for multi-cluster to get ServiceCIDR considering the proxyAll is disabled by default in most of cases.

@tnqn @wenyingd @hongliangl let us know if you know other alternative ways, thanks a lot!

hongliangl commented:
Or we can calculate Service CIDR only when mc is enable in AntreaProxy, and when proxyAll is enabled, we calculate Service CIDR and install the route for the CIDR.

jianjuns commented:
Yeah, I think it is good to keep a single way to discover Service CIDR. I am not sure what is the best design choice. Probably you guys can figure out.

luolanzone commented:
After sync with @tnqn and @hongliangl  we plan to extract the logic of serviceCIDR calculation out of AntreaProxy, and make it as standalone module, eg: servicecidr_discover.go. Both AntreaProxy and Multi-cluster can be a caller of new module. Hongliang will help to estimate the effort and submit a standalone PR. thanks.

wenyingd commented on pkg/agent/openflow/client.go:
```diff
@@ -618,11 +618,11 @@ func (c *client) GetPodFlowKeys(interfaceName string) []string {
 }
 
-func (c *client) InstallServiceGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints []proxy.Endpoint) error {
+func (c *client) InstallServiceGroup(groupID, exportedSvcGroupID binding.GroupIDType, withSessionAffinity bool, endpoints []proxy.Endpoint) error {
```
What is the expected behavior for a local Service deletion which is used as an Endpoint of the global Service?

luolanzone replied:
The local Service will be deleted, then corresponding Endpoint in the global Service will also be deleted as well. When the Endpoints are changed, InstallServiceGroup should be called again without `exportedSvcGroupID`.

luolanzone commented:
/test-multicluster-e2e

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -470,25 +482,33 @@ func (p *proxier) installServices() {
 				klog.ErrorS(err, "Error when installing Endpoints flows")
 				continue
 			}
+			var exportedSvcGroupID binding.GroupIDType
```
How about using a pointer here? Then you also don't need to use a dummy variable like `zeroGroupID`.

hongliangl replied:
Another question: for a mc-cluster Service, is it possible has more than one Endpoints which are all Services, if so, I think we need a slice to collect these exported Service group IDs, not a variable.

hongliangl replied:
> I don't get your question, for any exported Service, it will have only one corresponding local Service's ClusterIP as Endpoint of MC Service(IPv4 only).

Is it possible that a mc Service has multiple local Services as Endpoints?

luolanzone replied:
No, it's 1-1 mapping.

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -470,25 +482,33 @@ func (p *proxier) installServices() {
 			}
+			var exportedSvcGroupID binding.GroupIDType
+			if p.multiclusterEnabled && svcPortName.Name != mcWebhookServiceName && strings.HasPrefix(svcPortName.Name, "antrea-mc-") {
+				exportedSvcPortName := svcPortName
```
Maybe add some comments here since for developers that are not familiar with mc-cluster don't know why the Service name is got by removing prefix `antrea-mc-`.

hongliangl replied:
Maybe add some comments to explain why excluding Service `mcWebhookServiceName`?

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -470,25 +482,33 @@ func (p *proxier) installServices() {
+			if p.multiclusterEnabled && svcPortName.Name != mcWebhookServiceName && strings.HasPrefix(svcPortName.Name, "antrea-mc-") {
+				exportedSvcPortName := svcPortName
+				exportedSvcPortName.Name = strings.TrimPrefix(svcPortName.Name, "antrea-mc-")
+				if _, ok := p.serviceMap[exportedSvcPortName]; ok {
```
Why not handling the case that exported Service doesn't exist?

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -978,9 +1002,25 @@ func NewProxier(
 	return p
 }
 
+func registerServiceCIDREventHandler(p *proxier) {
```
I think this function could be a method of `proxier`.

jianjuns replied:
Even we reset in the test code, it can still be a func of proxier, unless we must pass proxier to the func (do we really need that?) in the test code?

luolanzone replied:
Done, re-based the code and refine the test part.

jianjuns commented on pkg/agent/controller/servicecidr/provider.go:
```diff
@@ -0,0 +1,150 @@
+	AddEventHandler(handler EventHandler)
+}
+
+type provider struct {
```
How about "discoverer" or "detector"?

luolanzone replied:
Sure, this part is from #4570 @hongliangl could you help to rename it? thanks.

hongliangl replied:
I'll rename that.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -978,9 +1013,24 @@ func NewProxier(
+	p.serviceCIDRProvider.AddEventHandler(func(serviceCIDRs []*net.IPNet) {
+		for _, cidr := range serviceCIDRs {
+			if cidr.IP.To4() != nil {
+				p.serviceCIDRIPv4 = cidr
```
Do we need lock? @hongliangl

luolanzone replied:
After sync with hongliang, I refined this part without event handler, proxy will get Service CIDR before sync rules.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -978,9 +1013,24 @@ func NewProxier(
 		p.endpointsConfig.RegisterEventHandler(p)
 	}
+	if p.multiclusterEnabled {
+		// We register event handler here since serviceCIDRIPv4 is only required
```
Probably remove "We register event handler here since", and just say: "serviceCIDRIPv4 is required only when Antrea Multi-cluster is enabled".

luolanzone replied:
Done

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -407,6 +417,8 @@ func (p *proxier) installServices() {
 		}
 
 		var allEndpointUpdateList, localEndpointUpdateList []k8sproxy.Endpoint
+		var isLocalServiceExported bool
```
Do we need this boolean, or we can just check "localService != nil"?

If we keep it, maybe name it "hasLocalServiceExported"

luolanzone replied:
Done, remove this variable.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -407,6 +417,8 @@ func (p *proxier) installServices() {
 
 		var allEndpointUpdateList, localEndpointUpdateList []k8sproxy.Endpoint
+		var isLocalServiceExported bool
+		var localService *agenttypes.LocalServiceInfo
```
Probably name it "mcsLocalService"

luolanzone replied:
Done

luolanzone commented on pkg/agent/types/multicluster.go:
```diff
@@ -0,0 +1,26 @@
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+)
+
+type LocalServiceInfo struct {
```
I was trying to define it in pkg/agent/openflow, but there is an import cycle issue. So I have to move it out.

luolanzone commented on pkg/agent/types/multicluster.go:
```diff
@@ -0,0 +1,26 @@
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+)
+
+type LocalServiceInfo struct {
```
Done, rename it as `ServiceGroupInfo` with comments.

luolanzone commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2457,12 +2459,19 @@ func (f *featureService) serviceEndpointGroup(groupID binding.GroupIDType, withS
-				Done()
+			// Antrea Multi-cluster support IPv4 only for now.
+			if localService != nil && endpointIP.Equal(localService.ClusterIP) {
+				group = group.Bucket().Weight(100).
```
Done

jianjuns commented on pkg/agent/types/multicluster.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+// ServiceGroupInfo is used in AntreaProxy for Antrea Multi-cluster.
+// It stores local exported Service's GroupID and its ClusterIP.
```
"a local exported"

jianjuns commented on pkg/agent/types/multicluster.go:
```diff
@@ -0,0 +1,30 @@
+	binding "antrea.io/antrea/pkg/ovs/openflow"
+)
+
+// ServiceGroupInfo is used in AntreaProxy for Antrea Multi-cluster.
```
Probably "for Multi-cluster Service load-balancing".

jianjuns commented on pkg/agent/types/multicluster.go:
```diff
@@ -0,0 +1,30 @@
+
+// ServiceGroupInfo is used in AntreaProxy for Antrea Multi-cluster.
+// It stores local exported Service's GroupID and its ClusterIP.
+type ServiceGroupInfo struct {
```
Why it causes cyclic import if we put `ServiceGroupInfo` in `pkg/agent/openflow`? I saw the type is used only in `pkg/agent/openflow` and `pkg/agent/proxy", and the latter already imports the former.

luolanzone replied:
The reason is the mock file in `pkg/agent/openflow` will also need `ServiceGroupInfo`, if we place it in `pkg/agent/openflow`, then the sub package will import a parent package.

jianjuns replied:
In this case, maybe we should add a "types" package under `pkg/agent/openflow`.

luolanzone replied:
Done, I will move all `agent/openflow` types into the new folder after this PR is merged considering there will be lots of import path change.

jianjuns replied:
Sure. No need to move all types. Just types and interfaces consumed by other packages.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -978,6 +1010,10 @@ func NewProxier(
 		p.endpointsConfig.RegisterEventHandler(p)
 	}
+	if p.multiclusterEnabled {
+		// serviceCIDRIPv4 is only required when Antrea Multi-cluster is enabled.
```
How about "serviceCIDRIPv4 is required only for Multi-cluster Services"?

luolanzone replied:
Done

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -416,6 +427,12 @@ func (p *proxier) installServices() {
 					needUpdateEndpoints = true
 				}
 			}
+			// When the Endpoint is local ClusterIP, it means corresponding local Service has been exported.
```
"a local ClusterIP"

jianjuns replied:
"the corresponding local Service is a member of the Multi-cluster Service."

luolanzone replied:
Done

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -470,25 +487,37 @@ func (p *proxier) installServices() {
+			// instead of a Multi-cluster Service, so it should be excluded. For any Multi-cluster Service, its name will
+			// be a combination with prefix `antrea-mc-` and exported Service's name. So we need to extract the prefix to
+			// find the exported Service if there is a match.
+			if p.multiclusterEnabled && mcsLocalService != nil && svcPortName.Name != mcWebhookServiceName &&
```
We should do this only when "mcsLocalService != nil"?

jianjuns replied:
Better to add func for "GetMCSExportedServiceName" or "GetMCSExportedServiceGroupID" to encapsulate the (tricky) logic.

luolanzone replied:
Yes, we need to check there is a matching local ClusterIP in the Endpoints first to make sure the Service is exported. So mcsLocalService won't be nil if there is a match.
Add two new functions to update `mcsLocalService`.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -470,25 +487,37 @@ func (p *proxier) installServices() {
 			}
+			// mcWebhookServiceName "antrea-mc-webhook-service" is a pre-defined webhook Service for Antrea Multi-cluster
+			// instead of a Multi-cluster Service, so it should be excluded. For any Multi-cluster Service, its name will
+			// be a combination with prefix `antrea-mc-` and exported Service's name. So we need to extract the prefix to
```
extract -> remove

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -470,25 +487,37 @@ func (p *proxier) installServices() {
+			// mcWebhookServiceName "antrea-mc-webhook-service" is a pre-defined webhook Service for Antrea Multi-cluster
+			// instead of a Multi-cluster Service, so it should be excluded. For any Multi-cluster Service, its name will
+			// be a combination with prefix `antrea-mc-` and exported Service's name. So we need to extract the prefix to
+			// find the exported Service if there is a match.
```
find -> look up

Probably remove "if there is a match"

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -346,6 +353,9 @@ func (p *proxier) uninstallLoadBalancerService(loadBalancerIPStrings []string, s
 
 func (p *proxier) installServices() {
+	if p.multiclusterEnabled {
+		p.serviceCIDRIPv4, _ = p.serviceCIDRProvider.GetServiceCIDR(false)
```
serviceCIDRIPv4 is used only in this func? Then why it needs to be a member of "proxier"?

luolanzone replied:
Done, define a local variable.

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +630,31 @@ func (p *proxier) installServices() {
 	}
 }
 
+func (p *proxier) GetMCSExportedServiceEndpointIP(serviceCIDRIPv4 *net.IPNet, endpointIP string) *agenttypes.ServiceGroupInfo {
```
Why capitalize the method? Is it used outside the package?

luolanzone replied:
Fixed, thanks.

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +630,31 @@ func (p *proxier) installServices() {
+	return mcsLocalService
+}
+
+func (p *proxier) GetMCSExportedServiceGroupID(mcsLocalService *agenttypes.ServiceGroupInfo, svcPortName k8sproxy.ServicePortName) {
```
Ditto

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +630,31 @@ func (p *proxier) installServices() {
+	// instead of a Multi-cluster Service, so it should be excluded. For any Multi-cluster Service, its name will
+	// be a combination with prefix `antrea-mc-` and exported Service's name. So we need to remove the prefix to
+	// look up the exported Service.
+	if mcsLocalService != nil && svcPortName.Name != mcWebhookServiceName && strings.HasPrefix(svcPortName.Name, "antrea-mc-") {
```
I think `mcsLocalService != nil` should be checked by the caller.

luolanzone replied:
Done

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +630,31 @@ func (p *proxier) installServices() {
+	// instead of a Multi-cluster Service, so it should be excluded. For any Multi-cluster Service, its name will
+	// be a combination with prefix `antrea-mc-` and exported Service's name. So we need to remove the prefix to
+	// look up the exported Service.
+	if mcsLocalService != nil && svcPortName.Name != mcWebhookServiceName && strings.HasPrefix(svcPortName.Name, "antrea-mc-") {
```
If we already checked `mcsLocalService != nil`, do we still need to check `svcPortName.Name != mcWebhookServiceName && strings.HasPrefix(svcPortName.Name, "antrea-mc-")`. At least, no need to check  `svcPortName.Name != mcWebhookServiceName`?

luolanzone replied:
You are right, refined this part.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +630,31 @@ func (p *proxier) installServices() {
+	// look up the exported Service.
+	if mcsLocalService != nil && svcPortName.Name != mcWebhookServiceName && strings.HasPrefix(svcPortName.Name, "antrea-mc-") {
+		exportedSvcPortName := svcPortName
+		exportedSvcPortName.Name = strings.TrimPrefix(svcPortName.Name, "antrea-mc-")
```
Please define a constant for "antrea-mc-" like multiclusterServiceNamePrefix.

luolanzone replied:
Done

luolanzone commented:
/test-multicluster-e2e

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +627,26 @@ func (p *proxier) installServices() {
+	}
+	// For any Multi-cluster Service, its name will be a combination with prefix `antrea-mc-` and
+	// exported Service's name. So we need to remove the prefix to look up the exported Service.
+	if mcsLocalService != nil && strings.HasPrefix(svcPortName.Name, mcServicePrefix) {
```
Should we return nil, if NOT `strings.HasPrefix(svcPortName.Name, mcServicePrefix)`, or NOT `p.serviceMap[exportedSvcPortName]`?

luolanzone replied:
Fixed.

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -50,6 +52,7 @@ const (
 	// SessionAffinity timeout is implemented using a hard_timeout in OVS. hard_timeout is
 	// represented by a uint16 in the OpenFlow protocol,
 	maxSupportedAffinityTimeout = math.MaxUint16
+	mcServicePrefix             = "antrea-mc-"
```
Name it `mcServiceNamePrefix` or add some comments to explain it is the prefix added to the MC Service name.

luolanzone replied:
Done

jianjuns commented on test/integration/ovs/ofctrl_test.go:
```diff
@@ -553,12 +553,12 @@ func TestBundleWithGroupAndFlow(t *testing.T) {
 	field3 := binding.NewRegField(3, 0, 31)
 	group := bridge.CreateGroup(groupID).
-		Bucket().Weight(100).
+		Bucket().Weight(openflow.DefaultWeight).
```
I feel no need to make openflow.DefaultWeight public just for usage here. But no strong opinion.

luolanzone replied:
Done

jianjuns commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -199,6 +200,9 @@ var (
 	// between the uplink and its pair port directly.
 	NonIPTable = newTable("NonIP", stageClassifier, pipelineNonIP, defaultDrop)
 
+	// Default bucket weight in a Openflow Group
```
a -> an

luolanzone commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -199,6 +200,9 @@ var (
 	NonIPTable = newTable("NonIP", stageClassifier, pipelineNonIP, defaultDrop)
 
+	// Default bucket weight in a Openflow Group
+	DefaultWeight = uint16(100)
```
Done

luolanzone commented:
/test-multicluster-e2e

jianjuns commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +628,31 @@ func (p *proxier) installServices() {
+	var groupID binding.GroupIDType
+	// For any Multi-cluster Service, its name will be a combination with prefix `antrea-mc-` and
+	// exported Service's name. So we need to remove the prefix to look up the exported Service.
+	if mcsLocalService != nil && strings.HasPrefix(svcPortName.Name, mcServiceNamePrefix) {
```
Questions:

1. If a local Service is exported after the MC Service is created, I assume the routine will still be triggered to update the MC Service's Group?
2. If the local Service does not have ready endpoints, its ClusterIP will not be added to the MCS Service's Endpoints?

luolanzone replied:
1. yes, a new Endpoint will be added to MC Service, so it will trigger and update MC Group.
2. correct, we only include the Service's ClusterIP when it has available Endpoints.

jianjuns reviewed (commented):
Two questions and two minor comments.

luolanzone commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +628,31 @@ func (p *proxier) installServices() {
+		exportedSvcPortName := svcPortName
+		exportedSvcPortName.Name = strings.TrimPrefix(svcPortName.Name, mcServiceNamePrefix)
+		if _, ok := p.serviceMap[exportedSvcPortName]; ok {
+			groupID = p.groupCounter.AllocateIfNotExist(exportedSvcPortName, false)
```
Done with first option. thanks.

luolanzone commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2442,7 +2446,9 @@ func (f *featureService) endpointDNATFlow(endpointIP net.IP, endpointPort uint16
 // will resubmit packets back to ServiceLBTable to trigger the learn flow, the learn flow will then send packets to
 // EndpointDNATTable. Otherwise, buckets will resubmit packets to EndpointDNATTable directly.
-func (f *featureService) serviceEndpointGroup(groupID binding.GroupIDType, withSessionAffinity bool, endpoints ...proxy.Endpoint) binding.Group {
+// When mcsLocalService is not nil, it means a local corresponding Service is exported,
```
Done

luolanzone commented:
/test-multicluster-e2e

jianjuns commented:
LGTM. Let us see if @wenyingd and @hongliangl have further comments.

wenyingd commented on pkg/agent/proxy/proxier.go:
```diff
@@ -476,19 +491,19 @@ func (p *proxier) installServices() {
 					// of the Service are different, install two groups. One group has all Endpoints, the other has only
 					// local Endpoints.
 					groupID := p.groupCounter.AllocateIfNotExist(svcPortName, true)
-					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, localEndpointUpdateList); err != nil {
```
Will NodePort/LoadBalancer type be used for a multi-cluster Service?

I ask it because I think when a local Service is working as an Endpoint in the multi-cluster Service, the Endpoint should not be a local Endpoint.

luolanzone replied:
No, we didn't support NodePort/LoadBalancer type in MC. But both NodePort/LoadBalancer have ClusterIP assigned, we may still treat them as ClusterIP according to KEP1645.

wenyingd commented on pkg/agent/proxy/proxier.go:
```diff
@@ -416,6 +428,9 @@ func (p *proxier) installServices() {
 				}
 			}
+			if p.multiclusterEnabled {
+				mcsLocalService = p.getMCSExportedServiceInfo(serviceCIDRIPv4, endpoint.IP(), svcPortName)
```
Is it possible that multiple local Services work as Endpoints in a single multi-cluster Service?

wenyingd replied:
Since we already find the updated local Service and its OF group ID, why do we still need to add the endpoint into `allEndpointUpdateList` ?

luolanzone replied:
1. No, the service name is unique, each Service will have its own multi-cluster Service.
2. If we didn't include it in allEndpointUpdateList, it will have a problem to update MC Service's Group when Service is no longer exported and Endpoint is removed, because there is no change in the cached installed Endpoint map.

luolanzone replied:
Done, refined the installed Endpoint map update logic.

wenyingd commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -2457,15 +2463,22 @@ func (f *featureService) serviceEndpointGroup(groupID binding.GroupIDType, withS
-				ResubmitToTable(resubmitTableID).
-				Done()
+			// Antrea Multi-cluster support IPv4 only for now.
+			if mcsLocalService != nil && endpointIP.Equal(mcsLocalService.ClusterIP) {
```
Why we have such scenario that `mcsLocalService != nil` but the endpoints does not contain "mcsLocalService.ClusterIP", and why not just set `mcsLocalService==nil` in that case?

luolanzone replied:
I don't get your point here, you mean to install normal Endpoint bucket with the check `mcsLocalService==nil`? This is in a loop of Endpoints list, there will be a case uncovered, we also need to install normal Endpoint bucket if `mcsLocalService != nil` and current Endpoint is not local Service's ClusterIP.

wenyingd replied:
My point is we should trust that the provided `mcsLocalService` and `endpoints` are desired to install in the OF Group, and no need to check the duplicated items in pkg/agent/openflow/pipeline.go . The caller of this function should guarantee all the parameters are valid.

luolanzone replied:
Done, pass the Endpoint without ClusterIP when mcsLocalService is not nil.

wenyingd commented on pkg/agent/proxy/proxier.go:
```diff
@@ -613,6 +628,27 @@ func (p *proxier) installServices() {
+		exportedSvcPortName := svcPortName
+		exportedSvcPortName.Name = strings.TrimPrefix(svcPortName.Name, mcServiceNamePrefix)
+		if _, ok := p.serviceMap[exportedSvcPortName]; ok {
+			mcsLocalService.GroupID = p.groupCounter.AllocateIfNotExist(exportedSvcPortName, false)
```
If all Endpoints of a local Service are removed but the Service is not removed yet, do we support but removing the bucket from the OF group which is used for the related multi-cluster Service?

luolanzone replied:
Yes, MC controller will be responsible to remove the Endpoint of MC Service if exported Service has no available Endpoints anymore. After the Endpoint is removed, the bucket will be removed because of Endpoint change.

hongliangl replied:
When the number of Endpoints of a Service is changed from N to 0, the group is still there and all buckets were removed.

wenyingd commented on pkg/agent/proxy/proxier.go:
```diff
@@ -476,19 +494,19 @@ func (p *proxier) installServices() {
 					// local Endpoints.
 					groupID := p.groupCounter.AllocateIfNotExist(svcPortName, true)
-					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, localEndpointUpdateList); err != nil {
+					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, nil, localEndpointUpdateList); err != nil {
```
Better to add a comment here to explain that a Service in multi-cluster scenario would not use NodePort/LoadBalancer type.

luolanzone replied:
Done.

wenyingd reviewed (commented):
LGTM overall, only one minor comment.

wenyingd reviewed (dismissed):
LGTM

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -476,19 +494,21 @@ func (p *proxier) installServices() {
-					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, localEndpointUpdateList); err != nil {
+					// The Multi-cluster Service support ClusterIP type of Service only, so always set mcsLocalService to nil when
+					// it's NodePort type of Service.
+					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, nil, localEndpointUpdateList); err != nil {
```
Is it possible that the ClusterIP of a NodePort/LoadBalancer as the Endpoint of a mcService?

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -346,6 +352,11 @@ func (p *proxier) uninstallLoadBalancerService(loadBalancerIPStrings []string, s
 }
 
 func (p *proxier) installServices() {
+	var serviceCIDRIPv4 *net.IPNet
```
If this is only used in getMCSExportedServiceInfo, why not getting this within getMCSExportedServiceInfo?

luolanzone replied:
I feel it's unnecessary to put this into the loop of handling `p.serviceMap`.

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -476,19 +494,21 @@ func (p *proxier) installServices() {
 					// local Endpoints.
 					groupID := p.groupCounter.AllocateIfNotExist(svcPortName, true)
-					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, localEndpointUpdateList); err != nil {
+					// The Multi-cluster Service support ClusterIP type of Service only, so always set mcsLocalService to nil when
```
```suggestion
					// The Multi-cluster Service supports ClusterIP type of Service only, so always set mcsLocalService to nil when
```

luolanzone replied:
Done

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -476,19 +494,21 @@ func (p *proxier) installServices() {
 					groupID := p.groupCounter.AllocateIfNotExist(svcPortName, true)
-					if err = p.ofClient.InstallServiceGroup(groupID, affinityTimeout != 0, localEndpointUpdateList); err != nil {
+					// The Multi-cluster Service support ClusterIP type of Service only, so always set mcsLocalService to nil when
+					// it's NodePort type of Service.
```
```suggestion
					// the type of the Service is not ClusterIP.
```

luolanzone replied:
Done

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
@wenyingd @hongliangl @jianjuns please let me know if you have further comments, we may move forward if no more comments, thanks.

wenyingd reviewed (dismissed):
LGTM

luolanzone commented:
@jianjuns could you help to take a look and move forward if there is no more comments? Thanks.

jianjuns commented:
/test-all

There are quite a few tests failed. Please make sure all tests passed.

luolanzone commented:
The kind tests are failed due to image download issue which is not related with the change, I triggered it again

luolanzone commented:
After checking the e2e log, I found it's a known codecov upload issue which is fixed by https://github.com/antrea-io/antrea/pull/4669, I will re-base and rerun.

luolanzone commented:
/test-all
/test-multicluster-e2e

jianjuns commented:
@luolanzone : you have a conflict to resolve.

luolanzone commented:
@jianjuns sure, I have addressed conflicts locally but I found there is a bug for Service flow installation when I tried to verify the changes in kind. It maybe introduced by a new change which is merged recently. @hongliangl is troubleshooting the bug, I will push the latest change after the bug is verified and fixed.

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
@jianjuns @hongliangl I have resolved conflicts, could you take a look again? thanks.

luolanzone commented:
/test-multicluster-e2e

jianjuns commented:
/test-all

hongliangl commented on pkg/agent/proxy/proxier.go:
```diff
@@ -48,8 +50,10 @@ const (
-	// represented by a uint16 in the OpenFlow protocol,
+	// represented by a uint16 in the OpenFlow protocol.
 	maxSupportedAffinityTimeout = math.MaxUint16
+	// The prefix added to the multi-cluster Service name.
```
```suggestion
	// The prefix of multi-cluster Service name.
```

hongliangl replied:
multi-cluster? Multi-cluster? Which should we use?

luolanzone replied:
when we talk about Antrea Multi-cluster feature, we will use `Multi-cluster`, if it's just referring to multi-cluster related concept, we will use `multi-cluster`.

hongliangl commented on pkg/agent/proxy/topology.go:
```diff
@@ -15,15 +15,24 @@
 )
 
-func (p *proxier) categorizeEndpoints(endpoints map[string]k8sproxy.Endpoint, svcInfo k8sproxy.ServicePort) ([]k8sproxy.Endpoint, []k8sproxy.Endpoint, []k8sproxy.Endpoint) {
+// When Antrea multi-cluster is enabled, the Endpoint from a local exported Service will be
```
reminder, multi-cluster/Multi-cluster

hongliangl commented on pkg/agent/proxy/topology.go:
```diff
@@ -15,15 +15,24 @@
-func (p *proxier) categorizeEndpoints(endpoints map[string]k8sproxy.Endpoint, svcInfo k8sproxy.ServicePort) ([]k8sproxy.Endpoint, []k8sproxy.Endpoint, []k8sproxy.Endpoint) {
+// When Antrea multi-cluster is enabled, the Endpoint from a local exported Service will be
+// represented as a ServiceGroupInfo instead of an element in the returned Endpoint slices.
+func (p *proxier) categorizeEndpoints(endpoints map[string]k8sproxy.Endpoint, svcInfo k8sproxy.ServicePort,
+	servicePortName k8sproxy.ServicePortName, serviceCIDRIPv4 *net.IPNet) ([]k8sproxy.Endpoint, []k8sproxy.Endpoint,
+	[]k8sproxy.Endpoint, *openflowtypes.ServiceGroupInfo) {
```
```suggestion
func (p *proxier) categorizeEndpoints(endpoints map[string]k8sproxy.Endpoint, 
svcInfo k8sproxy.ServicePort,
servicePortName k8sproxy.ServicePortName, 
serviceCIDRIPv4 *net.IPNet) ([]k8sproxy.Endpoint, 
[]k8sproxy.Endpoint,
[]k8sproxy.Endpoint, 
*openflowtypes.ServiceGroupInfo) {
```

hongliangl replied:
They should be either all in one line or per parameter in one line.

luolanzone replied:
Done

hongliangl commented on pkg/agent/proxy/topology.go:
```diff
@@ -35,6 +44,13 @@ func (p *proxier) categorizeEndpoints(endpoints map[string]k8sproxy.Endpoint, sv
 			if useTopology && !availableForTopology(ep, p.nodeLabels) {
 				return false
 			}
+			if p.multiclusterEnabled {
+				tempLocalService := p.getMCSExportedServiceInfo(serviceCIDRIPv4, ep, servicePortName)
+				if tempLocalService != nil {
+					mcsLocalService = tempLocalService
+					return false
+				}
+			}
```
```suggestion
                        if p.multiclusterEnabled {
				tempLocalService := p.getMCSExportedServiceInfo(serviceCIDRIPv4, ep, servicePortName)
				if tempLocalService != nil {
					mcsLocalService = tempLocalService
					return false
				}
			}
			if useTopology && !availableForTopology(ep, p.nodeLabels) {
				return false
			}
```

hongliangl replied:
When ToplologyAwareHints is enabled, if the special Multi-Cluster Endpoint is filter out, mcsLocalService will never be updated.

luolanzone replied:
Done

luolanzone commented:
Seems the multicluster e2e take longer time then before and a few MCNP cases are failed randomly. I tried a failed case locally, it works fine. syncing with @hjiajing and @GraysonWu regarding the failed issue.

luolanzone commented:
/test-multicluster-e2e

luolanzone commented:
The MC testbed is recovered and e2e test is passed.

luolanzone commented:
/test-all

hongliangl reviewed (approved):
LGTM

tnqn commented on pkg/agent/proxy/proxier.go:
```diff
@@ -403,14 +414,18 @@ func (p *proxier) installServices() {
 				break
 			}
 		}
+		if mcsLocalService != nil {
+			needUpdateEndpoints = true
+		}
```
This means the whole endpoints of this service will be updated as long as there is any service/endpoints change in the whole cluster and when the peridiocal sync happens (30s).

Why it must update endpoints when there is a local service endpoint?

tnqn commented on pkg/agent/proxy/topology.go:
```diff
@@ -152,14 +172,36 @@ func (p *proxier) canUseTopology(endpoints map[string]k8sproxy.Endpoint, svcInfo
+			Endpoint: endpoint,
+		}
+	}
+	// For any Multi-cluster Service, its name will be a combination with prefix `antrea-mc-` and
+	// exported Service's name. So we need to remove the prefix to look up the exported Service.
+	if mcsLocalService != nil && strings.HasPrefix(svcPortName.Name, mcServiceNamePrefix) {
+		exportedSvcPortName := svcPortName
+		exportedSvcPortName.Name = strings.TrimPrefix(svcPortName.Name, mcServiceNamePrefix)
+		if _, ok := p.serviceMap[exportedSvcPortName]; ok {
+			mcsLocalService.GroupID = p.groupCounter.AllocateIfNotExist(exportedSvcPortName, false)
+			return mcsLocalService
+		}
+	}
+	return nil
```
this looks a little hacky, making AntreaProxy too specific to Multicluster. It would be hard to maintain in long term if we add such logic to it.
And I think the way doesn't always work as even it could allocate an ID for this service, it may not exist in OVS and installing flows with this group will fail.
I wonder if we could have more generic code to handle cluster IP as endpoint IP's case, not specific to Multicluster.

jianjuns replied:
If flow installation will fail when the group does not exist, then we need to wait for another 30s for retry?

No good idea about decoupling AntreaProxy and Multicluster.

tnqn replied:
For decoupling AntreaProxy and Multicluster, I discussed this with @luolanzone offline. The proposal is:
1. Add a attribute `IsNested` to `ServiceInfo` to indicate whether the service may be DNATed more than one.
```
// ServiceInfo is the internal struct for caching service information.
type ServiceInfo struct {
	*k8sproxy.BaseServiceInfo
	// cache for performance
	OFProtocol openflow.Protocol
 +      // InNested represents the Service's Endpoints could be another Service's ClusterIP.
 +      IsNested bool
}

IsNested = service.Annotations[MulticlusterSpecificAnnotation]
```

2. For a nested Service, its Service flow will mark one reg bit of the packet indicate "nested":
```
InstallServiceFlows(groupID binding.GroupIDType, 
        svcIP net.IP, 
        svcPort uint16, 
        protocol binding.Protocol, 
        affinityTimeout uint16, 
        nodeLocalExternal bool, 
+       nested bool,
        svcType v1.ServiceType) 
```

3. For a non nested Service, it will install one extra flow in EndpointDNAT table to resubmit packets with "nested" bit to its own group with higher priority.

The order of installing multicluster-service and normal service doesn't matter as they have their flows are indidivual. When multicluster-service's local service is not installed, the packet will just be DNATed to local service IP as normal packet
When multicluster-service's local service is not installed, the packet will be resubmitted to the local service's group to select the eventual endpoint.

The flows would look like the below, assuming 10.96.0.1 as multicluster service IP and 10.96.0.2 as normal service IP:
```
table=ServiceLB, priority=200,tcp,nw_dst=10.96.0.1,tp_dst=80 actions=load:0x1->NESTED-BIT,group:1 
table=ServiceLB, priority=200,tcp,nw_dst=10.96.0.2,tp_dst=80 actions=group:2
table=EndpointDNAT, priority=210,tcp,NESTED-BIT=0x1,reg3=0xa060002,reg4=0x20050/0x7ffff actions=group:2
table=EndpointDNAT, priority=200,tcp,reg3=0xa060002,reg4=0x20050/0x7ffff actions=ct(commit,table=AntreaPolicyEgressRule,zone=65520,nat(dst=10.96.0.2:80)
```

Whether to do all of these could be managed by a flag in AntreaProxy like supportedNestedService and it will be assigned the value of enableMulticlusterGateway, but the approach could potentially supports in-cluster nested Service in the future. What do you think?

tnqn reviewed (commented):
Sorry for the late review. I found conflict with my another PR and took a quick look, have some questions below.
